packages/app/contexts/TokenListContext.tsx

```diff
@@ -144,42 +129,56 @@ export const TokenListProvider = ({ children }: PropsWithChildren) => {
   }, [address, callArray, chainId, tokenList]);
 
   const setupTokenList = useCallback(async () => {
-    let mergedTokenlistTokens = defaultTokenList;
     setIsLoading(true);
     async function getTokenListData(tokenlistURL: string) {
-      const res = await fetch(tokenlistURL);
+      const { signal } = abortControllerRef.current;
+      const res = await fetch(tokenlistURL, { signal });
       if (!res.ok) {
         throw new Error("Failed to fetch tokenlist data");
       }
       return res.json();
     }
 
-    Promise.allSettled(
+    await Promise.allSettled(
       fetchTokenlistUrls.map((list) => getTokenListData(list))
     ).then((results) => {
       results.forEach((result) => {
         if (result.status === "fulfilled") {
-          mergedTokenlistTokens = mergeTokenlists(
-            mergedTokenlistTokens,
-            result.value.tokens
-          );
+          const mergedTokenlist = [...defaultTokenList, ...result.value.tokens];
+
+          const tokenUniqueAddressesSet = new Set<string>();
+          const tokenListNoDuplicates = mergedTokenlist.filter((token) => {
+            if (
+              token.chainId === chainId &&
+              !tokenUniqueAddressesSet.has(token.address.toLowerCase())
+            ) {
+              tokenUniqueAddressesSet.add(token.address.toLowerCase());
+              return true;
+            }
+            return false;
+          });
+
+          setTokenList(tokenListNoDuplicates);
+          setIsLoading(false);
         } else {
           console.error("Error fetching tokenlist data:", result.reason);
         }
       });
-
-      setTokenList(
-        mergedTokenlistTokens.filter(
-          (token: TokenFromTokenlist) => token.chainId === chainId
-        )
-      );
-      setIsLoading(false);
     });
   }, [chainId, defaultTokenList, fetchTokenlistUrls]);
 
   useEffect(() => {
+    // abort if still fetching other chain async
+    abortControllerRef.current.abort();
+    abortControllerRef.current = new AbortController();
     setupTokenList();
-  }, [address, chainId, setupTokenList]);
+
+    return () => {
+      // Cleanup: Abort ongoing fetch when component unmounts or chainId changes
```

Comment:
I think this Cleanup comment isn't accurate. This doesn't trigger when chainId changes

Reply:
I'll remove it